### PR TITLE
Add support for Crypto

### LINKED ISSUE #5019: DOM missing defs for crypto
There is node crypto, but no browser crypto in the defs.

### PR DESCRIPTION
**Pull request by MichaelDeBoey**
Closes #5019

Ref: https://developer.mozilla.org/en-US/docs/Web/API/Crypto

### CODE REVIEW

File: lib/dom.js [hunk 1]
```diff
@@ -4,6 +4,117 @@
...
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW';
+declare interface SubtleCrypto {
+  decrypt(algorithm: SubtleCrypto$Algorithm, key: CryptoKey, data: BufferSource): Promise<string>;
+  // TODO: fix type for 'algorithm' param
+  deriveBits(algorithm: any, baseKey: CryptoKey, length: number): Promise<ArrayBuffer>;
+  deriveKey(algorithm: SubtleCrypto$DeriveKeyAlgorithm, masterKey: CryptoKey, derivedKeyAlgorithm: DOMString | SubtleCrypto$DerivedKeyAlgorithm, extractable: Boolean, keyUsages: CryptoKey$Usages[]): Promise<CryptoKey | CryptoKeyPair>;
+  digest(algo: SubtleCrypto$HashAlgo, buffer: BufferSource): Promise<string>;
```
Comment (qm3ster) [inline lib/dom.js]: This definitely returns something `ArrayBuffer`-like and not a string.

File: lib/dom.js [hunk 2]
```diff
@@ -4,6 +4,117 @@
...
+  encrypt(algorithm: SubtleCrypto$Algorithm, key: CryptoKey, data: BufferSource): Promise<ArrayBuffer>;
+  exportKey(format: SubtleCrypto$KeyFormat, key: CryptoKey): Promise<CryptoKey>;
+  generateKey(algo: SubtleCrypto$GenerateKeyAlgo, extractable: Boolean, keyUsages: CryptoKey$Usages[]): Promise<CryptoKey | CryptoKeyPair>;
+  importKey(format: SubtleCrypto$KeyFormat, keyData: ArrayBuffer | JSONWebKey, algo: SubtleCrypto$GenerateKeyAlgo, extractable: Boolean, usages: CryptoKey$Usages[]): Promise<CryptoKey>;
+  sign(algo: SubtleCrypto$SignAlgo, key: CryptoKey, text2sign: BufferSource): Promise<BufferSource>;
```
Comment (qm3ster) [inline lib/dom.js]: How did you choose to return a `BufferSource`-implementing type rather than just an `ArrayBuffer`? both **MDN** and **w3.org** itself fail to specify the content of the promise, at the time of writing. Could it depend on the algo?(unlikely)

File: lib/dom.js [hunk 3]
```diff
@@ -4,6 +4,117 @@
...
+  | 'RSASSA-PKCS1-v1_5'
+  | 'ECDSA';
+type SubtleCrypto$WrapAlgo =
+  | 'AES-CBC'
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW';
+declare interface SubtleCrypto {
+  decrypt(algorithm: SubtleCrypto$Algorithm, key: CryptoKey, data: BufferSource): Promise<string>;
```
Comment (qm3ster) [inline lib/dom.js]: Also definitely `Promise<ArrayBuffer>`, the example uses TextDecoder and I looked up type in Chrome and FF debugger.
```js
(async function test(){

const ptUtf8 = new TextEncoder().encode('my plaintext')

const pwUtf8 = new TextEncoder().encode('my password')
const pwHash = await crypto.subtle.digest('SHA-256', pwUtf8)

const iv = crypto.getRandomValues(new Uint8Array(12))
const alg = { name: 'AES-GCM', iv: iv }
const key = await crypto.subtle.importKey('raw', pwHash, alg, false, ['encrypt'])

const ctBuffer = await crypto.subtle.encrypt(alg, key, ptUtf8)

const key1 = await crypto.subtle.importKey('raw', pwHash, alg, false, ['decrypt'])

const ptBuffer = await crypto.subtle.decrypt(alg, key1, ctBuffer)

const plaintext = new TextDecoder().decode(ptBuffer)
debugger
return plaintext
}())
```

File: lib/dom.js [hunk 4]
```diff
@@ -4,6 +4,117 @@
...
+
+// This is a helper type to simplify the specification, it isn't an interface
+// and there are no objects implementing it.
+// https://developer.mozilla.org/en-US/docs/Web/API/BufferSource
+type BufferSource = ArrayBuffer | $ArrayBufferView;
+
+type DOMString = string;
+
+declare interface Crypto extends RandomSource {
+  subtle: SubtleCrypto; // readonly
```
Comment (qm3ster) [inline lib/dom.js]: My head is currently in TypeScript land, but I am almost certain [this is the way to make read-only props on an interface](https://flow.org/en/docs/types/interfaces/#toc-interface-property-variance-read-only-and-write-only).

File: lib/dom.js (hunk 2 above)
Comment (MichaelDeBoey) [reply lib/dom.js]: Because the input (`text2sign `) is of type `BufferSource`, it seemed logical the `Promise` would be something of `BufferSource` as well

Comment (MichaelDeBoey) [reply lib/dom.js]: Because the input (`text2sign `) is of type `BufferSource`, it seemed logical the `Promise` would be something of `BufferSource` as well

File: lib/dom.js (hunk 1 above)
Comment (MichaelDeBoey) [reply lib/dom.js]: This is was not defined in any spec, but it should be a buffer indeed 🙂
`BufferSource` is more appropriate I think, since the input (`buffer`) is of type `BufferSource`?

File: lib/dom.js (hunk 3 above)
Comment (MichaelDeBoey) [reply lib/dom.js]: `BufferSource` is more appropriate I think, since the input (`data `) is of type `BufferSource`?

File: lib/dom.js [hunk 5]
```diff
@@ -4,6 +4,117 @@
...
+// TODO: CryptoKeyPair
+declare interface CryptoKeyPair {
+
+}
+
+// TODO: JSONWebKey
+type JSONWebKey = {};
+
+type RandomSource =  {
+  getRandomValues(typedArray: $TypedArray): void;
```
Comment (qm3ster) [inline lib/dom.js]: The [MDN article](https://developer.mozilla.org/en-US/docs/Web/API/RandomSource/getRandomValues) uses mutating the array, but [w3.org](https://www.w3.org/TR/WebCryptoAPI/#Crypto-method-getRandomValues) officially specifies the array must also be returned.
MDN also depends on that feature in [the `encrypt` and `decrypt` examples](https://github.com/facebook/flow/pull/5027/commits/d56ffe786f3509999c304605e36e96979f04a3bd#r145700797).

TL;DR: I imagine it should go a little something like `getRandomValues<T: $TypedArray>(typedArray: T): T;`

File: lib/dom.js (hunk 4 above)
Comment (MichaelDeBoey) [reply lib/dom.js]: That's indeed the way to make read-only props on an interface, but I did it this way, because in other parts of the file it was also done like this and I wasn't sure what to do with them so I stuck to the current way

File: lib/dom.js (hunk 3 above)
Comment (qm3ster) [reply lib/dom.js]: It doesn't preserve the type in any way though? It's impossible to guess what was used from the ciphertext, so it's a safe bet it will always be an `ArrayBuffer`? :thinking:
It literally operates on the raw bytes.

File: lib/dom.js (hunk 5 above)
Comment (MichaelDeBoey) [reply lib/dom.js]: I just followed the [MDN spec](https://developer.mozilla.org/en-US/docs/Web/API/RandomSource/getRandomValues), but if [w3.org](https://www.w3.org/TR/WebCryptoAPI/#Crypto-method-getRandomValues) specifies it with a return-variable and MDN uses it in there examples, this one should defo have one too. 🙂

File: lib/dom.js (hunk 1 above)
Comment (qm3ster) [reply lib/dom.js]: Regarding all methods:
Saying it is a `BufferSource` won't break anything, but I don't see why any vendor would return an array representation of a buffer with a particular bitness and not a naked buffer.
I suspect `ArrayBuffer` is more correct for the returns.
`BufferSource` is indeed the type for all inputs, even `decrypt`.

Comment (MichaelDeBoey) [reply lib/dom.js]: OK will make them `ArrayBuffer` for now 🙂
We can always change them to `BufferSource` if necessary 🙂

**Comment by MichaelDeBoey** [comment]
@qm3ster I addressed your comments 🙂

**Comment by MichaelDeBoey** [comment]
@qm3ster I addressed your comments 🙂

The CI tests will still fail, 'cause of the line-numbers.
Don't know what's the best way to make them pass?

File: lib/dom.js [hunk 6]
```diff
@@ -4,6 +4,116 @@
...
+
+// TODO: CryptoKeyPair
+declare interface CryptoKeyPair {
+}
+
+// TODO: JSONWebKey
+type JSONWebKey = {};
+
+type RandomSource = {
+    getRandomValues(typedArray: $TypedArray): $TypedArray;
```
Comment (qm3ster) [inline lib/dom.js]: This should be a generic, because it is always guaranteed to return the same member of $TypedArray, since it just returns the same instance.

Comment (qm3ster) [reply lib/dom.js]: Is there any reason `getRandomValues<T: $TypedArray>(typedArray: T): T;` didn't work?

**Comment by qm3ster** [comment]
How naughty to do an interactive rebase on a publically accessible branch :rofl:

**Comment by MichaelDeBoey** [comment]
@qm3ster I know, but it makes the commit cleaner, which would help when merging 🙂

**Comment by MichaelDeBoey** [comment]
@qm3ster The CI tests will still fail, 'cause of the line-numbers.
Don't know what's the best way to make them pass?

**Comment by qm3ster** [comment]
@MichaelDeBoey that it do, but it hid the old CI reference so I couldn't see why it fails + it disconnected my review comments from the code so I had to go line hunting.
I think it's best to do that at the end, before the merge.
Anyway, I am happy with the result and will show myself out.

**Comment by MichaelDeBoey** [comment]
@asolove Maybe you can point me in the right direction on what's the best way to make the failed CI tests ('cause of the line-numbers) pass?

**Comment by qm3ster** [comment]
Well just to make it pass as is, overwriting `tests/dom/dom.exp` would be enough.

However, this brings to my attention: 
does this need any tests?
or are they only for preventing regressions on real bugfixes?

**Comment by MichaelDeBoey** [comment]
@qm3ster Indeed changing the line-numbers in `tests/dom/dom.exp` would fix the tests, but it looks very cumbersome to do so every time we add/update something in `lib/dom.js` (or other files, which will cause failures in other tests), so that's why I was asking for a better solution.

**Comment by qm3ster** [comment]
Disabling line numbers would make it easier to see the change in the diff when test results change, and avoid failing changes to the lib when tests and test results are unchanged. But that is still a hack.
The everchanging libs should not be used as an integration test. There should be errors expected inside each test with the javascript API. The error printer should not be involved.

**Comment by MichaelDeBoey** [comment]
Can somebody point me in the right direction on what's the best way to make the failed CI tests ('cause of the line-numbers) pass please?

CC: @asolove @samwgoldman @mroch @nmote

**Comment by qm3ster** [comment]
Maybe @asolove is kill 🤔

**Comment by MichaelDeBoey** [comment]
@nmote Can take a look at it later today if that's OK for you guys?

**Comment by MichaelDeBoey** [comment]
@qm3ster @nmote I created #7714,  since I deleted the repo from my account
